fix(dashboard): improve experience for screen readers in cumulative tests by state section (small multiples)

nselikoff opened the pull request:
Specifically, this PR hides the small multiples and corresponding legend to screen readers until we can improve the accessibility of the AreaChart component further. Also adds "Sort by: " text to add context to the toggle between total and per-capita, which will clarify to all users what the toggle does.

Note that by default, if focusable and ariaHidden are not supplied to the AreaChart component, they will not be rendered in the HTML (i.e. both are "opt in").

Also note there is a visual change with the addition of "Sort by":

<img width="477" alt="Screen Shot 2020-04-16 at 10 17 01 PM" src="https://user-images.githubusercontent.com/1994207/79525082-0a6e2e80-8030-11ea-82b2-d9a57c86df95.png">

Fixes #610

goleary commented:
I like what you've done here, but I wonder about the 'sort by' language.

That toggle not only changes the sort property, but also the display properties and "Sort by" doesn't feel like it fully capture that.

jeremiak commented:
I agree. What about some of these options (though I think in all cases we should either just sentence case the option or entirely lower case as if its part of the sentence:

* "Look at"
* "Display by"
* "Look at charts based on"

nselikoff commented:
@goleary @jeremiak great call. I like "Display by" personally, just added that and changed to sentence case in the latest commit:

![image](https://user-images.githubusercontent.com/1994207/79614766-3c829d80-80cf-11ea-92e1-4ed19015de19.png)

vijaysharwar commented on src/pages/dashboard/_StateCumulativeTestsContainer.js:
```diff
@@ -191,20 +191,25 @@ export default function CumulativeTestsByStateContainer() {
-            Tests Per Capita
-          </span>
+        <div className="dashboard-toggle-wrapper">
+          <div className="dashboard-toggle-label">Display by:</div>
```
This layout breaks on small screens. This is at iPhone X dimensions:

<img width="374" alt="Screen Shot 2020-04-18 at 11 27 52 AM" src="https://user-images.githubusercontent.com/63169602/79641811-c2eabe00-8167-11ea-8fbf-a0851f24cc85.png">

nselikoff replied:
@vijaysharwar thank you, just pushed an update for this (below screenshot at iPhone X dimensions):

![image](https://user-images.githubusercontent.com/1994207/79642446-7ef9b800-816b-11ea-97b2-7bb3a47c714f.png)

vijaysharwar commented:
@nselikoff curious about rationale behind leaving some other surrounding elements.

https://github.com/COVID19Tracking/website/blob/e1398d1fc1f0ebd45e892643f07d1f5fffb94461/src/pages/dashboard/_StateCumulativeTestsContainer.js#L183-L211

Especially the toggle. Seems like if the charts are aria-hidden, then it doesn't make sense to have a screen reader detect those controls (there's nothing actionable for the user).

nselikoff commented:
@vijaysharwar good question - since the toggle changes the order of the listing of the states (and the state links will still be present to screen readers), my thought was that leaving it could still preserve some of the intended meaning (i.e. which states are "leading the race" in terms of testing when seen through those two lenses).

vijaysharwar commented:
@nselikoff ah, that makes sense. So the titles are still displaying. In that case, a nice improvement might be to put the state multiples into an `<ol>` instead of `<div>` (suppressing the numbers with CSS) so that screen readers denote it as an ordered/ranked list.

And I think the SVGs could probably be made fairly screen reader accessible by proving ARIA labels. Like, maybe we could have a screen reader only text intro for each state's chart with the most recent totals/cumulative data.

Further reading/ideas: https://www.a11ywithlindsey.com/blog/accessibility-d3-bar-charts

nselikoff commented:
@vijaysharwar I like that idea re: `<ol>`. 

Also agree that an ARIA label for the small multiples would be better, and not difficult to highlight the most recent 2 data points + the stay at home annotation if it exists.

I'm wondering if the toggle should be marked as `role="radiogroup"` instead of switch (https://www.w3.org/TR/wai-aria-practices-1.1/#radiobutton) since we're really toggling between two interpretations, not switching a state on and off. Thoughts on that?

When I was reading about data viz accessibility the other day, in addition to the great link you shared, I also came across a good example from HighCharts of a line chart with a11y-focused construction, annotation and markup: https://www.highcharts.com/samples/highcharts/accessibility/accessible-line/. After v1 launches I think we should have a larger discussion on Slack about what we could do to further improve the AreaChart and other components.

vijaysharwar commented:
@nselikoff yeah, I've been curious about rationale behind `role="switch"` myself. It's not something I've used often, would need to read up on how various screen readers/browsers handle it. I don't know off top of my head if `"radiogroup"` would perform better.

However, there are as always lots of ways to skin this cat. I've been wondering if a [tab navigation](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role) with `role="tab"` and `aria-controls` attributes might be best. That way, you can link the toggle button to the new content that is rendered.

None of this should hold up V1 launch, IMO. Probably worth merging this, and starting a new issue where we can define improvements over GItHub and Slack.
